rust/capture/src/server.rs

```diff
@@ -120,26 +122,57 @@ where
                 Some(partition)
             }
         };
-        let sink = KafkaSink::new(
+        let kafka_sink = KafkaSink::new(
             config.kafka,
             sink_liveness,
             partition,
             replay_overflow_limiter,
         )
         .expect("failed to start Kafka sink");
 
-        router::router(
-            crate::time::SystemTime {},
-            liveness,
-            sink,
-            redis_client,
-            billing_limiter,
-            token_dropper,
-            config.export_prometheus,
-            config.capture_mode,
-            config.concurrency_limit,
-            event_max_bytes,
-        )
+        if config.s3_fallback_enabled {
+            let sink_liveness = liveness
+                .register("s3".to_string(), Duration::seconds(30))
+                .await;
+
+            let s3_sink = S3Sink::new(
+                config
+                    .s3_fallback_bucket
+                    .expect("S3 bucket required when fallback enabled"),
+                config.s3_fallback_endpoint,
+                sink_liveness,
+            )
+            .await
+            .expect("failed to create S3 sink");
+
+            let fallback_sink = FallbackSink::new(kafka_sink, s3_sink);
+
+            router::router(
+                crate::time::SystemTime {},
+                liveness,
+                fallback_sink,
+                redis_client,
+                billing_limiter,
+                token_dropper,
+                config.export_prometheus,
+                config.capture_mode,
+                config.concurrency_limit,
+                event_max_bytes,
+            )
+        } else {
+            router::router(
+                crate::time::SystemTime {},
+                liveness,
+                kafka_sink,
+                redis_client,
+                billing_limiter,
+                token_dropper,
+                config.export_prometheus,
+                config.capture_mode,
+                config.concurrency_limit,
+                event_max_bytes,
+            )
+        }
```

Comment:
Feels a tad over-duplicated. Could we not assign the sink to a var and then just do the router:: stuff at the end?

Reply:
Yeah, I tried for ages to do that then gave up

Would need to have sink be type Box<dyn Events + Send + Sink> and then need to change router and everything to use boxed stuff and in the end felt less boilerplate to just duplicate the router creation

Definitely my least favourite part of rust the stupid known size stuff

Reply:
there you go: https://github.com/PostHog/posthog/pull/28042/commits/570d2fac55bd35cd5b812d571bf18ce25eac9088